c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,6 +224,49 @@ struct Block {
   }
 };
 
+// BlockState, BlockPoolState, and PrivatePoolState contain the information
+// needed to reconstruct a private pool to a previous state. See note
+// [Checkpointing PrivatePoolState]
+struct BlockState {
+  int device = 0;
+  cudaStream_t stream = 0;
+  stream_set stream_uses = {};
+  size_t size = 0;
+  void* ptr = nullptr;
+  bool allocated = false;
+  int gc_count = 0;
+  // maintain invariant that event_count == 0 ;
+  BlockState* prev = nullptr;
+  BlockState* next = nullptr;
+  // history will be left alone in checkpoint
+
+  BlockState(Block* block);
+};
+
+struct SegmentState {
+  std::vector<std::shared_ptr<BlockState>> blocks;
```

Comment:
This can just be vector<BlockState> if we remove the prev/next pointers (see other comments).